src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -554,36 +556,89 @@ private void CheckForHttp11ConnectionInjection()
+                    Debug.Assert(_associatedHttp11ConnectionCount >= _pendingHttp11ConnectionCount,
+                        $"Expected {_associatedHttp11ConnectionCount} >= {_pendingHttp11ConnectionCount}");
 
-                    int availableConnectionCount = _availableHttp11Connections.Count;
-                    if (availableConnectionCount > 0)
+                    if (_http11RequestQueue.Count != 0)
                     {
-                        // We have a connection that we can attempt to use.
-                        // Validate it below outside the lock, to avoid doing expensive operations while holding the lock.
-                        connection = _availableHttp11Connections[availableConnectionCount - 1];
-                        _availableHttp11Connections.RemoveAt(availableConnectionCount - 1);
+                        if (connection is not null || _http11Connections.TryDequeue(out connection))
+                        {
+                            bool success = _http11RequestQueue.TryDequeueWaiter(this, out waiter);
+                            Debug.Assert(success == waiter is not null);
+                        }
                     }
-                    else
+
+                    // Update the empty queue flag now.
+                    _http11RequestQueueIsEmptyAndNotDisposed = _http11RequestQueue.Count == 0 && !_disposed;
+
+                    if (waiter is null)
                     {
-                        // No available connections. Add to the request queue.
-                        waiter = _http11RequestQueue.EnqueueRequest(request);
+                        // We didn't find a waiter to signal, or there were no connections available.
+
+                        if (connection is not null)
+                        {
+                            // A connection was provided to this method, or we rented one from the pool.
+                            // Return it back to the pool since we're not going to use it yet.
 
-                        CheckForHttp11ConnectionInjection();
+                            // We're returning it while holding the lock to avoid a scenario where
+                            // - thread A sees no requests are waiting in the queue (current thread)
+                            // - thread B adds a request to the queue, and sees no connections are available
+                            // - thread A returns the connection to the pool
+                            // We'd have both a connection and a request waiting in the pool, but nothing to pair the two.
 
-                        // There were no available idle connections. This request has been added to the request queue.
-                        if (NetEventSource.Log.IsEnabled()) Trace($"No available HTTP/1.1 connections; request queued.");
-                        connection = null;
-                        return false;
+                            // This should be a rare case, so the added contention should be minimal.
```

Comment:
It should be rare that we're returning a connection to the pool and there are no pending requests? That doesn't seem like it would actually be rare, but it would seem logical that the added contention in that case wouldn't be very impactful, since if no one's waiting, either there isn't much demand or everyone was already satisfied.

Reply:
It should be rare that we saw `queueIsEmpty == false` to get to this method, followed by checking and it actually being empty

Reply:
So the intent is this method is only called when the (request?) queue was observed to be empty? Can you add such assumptions to an XML comment on the method?

Reply:
This would be called when:
- Returning a connection and observing that the request queue is not empty
- After adding a request to the queue when failing to pick a connection from the queue
- after scavenging or disposing to deal with race conditions

For the first case, the request queue is likely to not be empty by the time we take the lock. But there's a chance it could be (multiple returning connections all saw the flag being set / a request was cancelled at just the right time), in which case we would have to return it in this branch.
This one may happen in small spikes when we run out of connections before we're able to inject more.

In the second case,  we could end up renting a connection but then have to return it if TryDequeue returned false (e.g. due to cancellation just happening for the remainder of requests). This should be very rare.

The third one doesn't matter since it's so rare relative to other queue operations.


I'll add a large comment at the top